src/main/java/com/gtnewhorizons/modularui/api/math/MathExpression.java

```diff
@@ -1,315 +1,593 @@
         }
 
-        for (int i = 1; i < parsed.size() - 1; i++) {
-            Object obj = parsed.get(i);
-            if (obj == Operator.SCIENTIFIC) {
-                Double left = (Double) parsed.get(i - 1);
-                Double right = (Double) parsed.get(i + 1);
-                Double result = left * Math.pow(10, right);
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.add(i - 1, result);
-                i--;
-            }
-        }
+        // Strip all spaces from the input string.
+        // This also correctly interprets numbers in the French locale typed by user using spaces as thousands
+        // separators.
+        // See: https://bugs.java.com/bugdatabase/view_bug?bug_id=4510618
+        expr = expr.replace(" ", "");
 
-        // ^ is right-associative: a^b^c = a^(b^c)
-        for (int i = parsed.size() - 2; i > 0; i--) {
-            Object obj = parsed.get(i);
-            if (obj == Operator.POWER) {
-                Double left = (Double) parsed.get(i - 1);
-                Double right = (Double) parsed.get(i + 1);
-                Double result = Math.pow(left, right);
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.add(i - 1, result);
-                i--;
-            }
+        if (expr.isEmpty()) {
+            ctx.success = true;
+            return ctx.emptyValue;
         }
 
-        for (int i = 1; i < parsed.size() - 1; i++) {
-            Object obj = parsed.get(i);
-            if (obj == Operator.MULTIPLY) {
-                Double left = (Double) parsed.get(i - 1);
-                Double right = (Double) parsed.get(i + 1);
-                Double result = left * right;
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.add(i - 1, result);
-                i--;
-            } else if (obj == Operator.DIVIDE) {
-                Double left = (Double) parsed.get(i - 1);
-                Double right = (Double) parsed.get(i + 1);
-                Double result = left / right;
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.add(i - 1, result);
-                i--;
-            } else if (obj == Operator.MOD) {
-                Double left = (Double) parsed.get(i - 1);
-                Double right = (Double) parsed.get(i + 1);
-                Double result = left % right;
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.remove(i - 1);
-                parsed.add(i - 1, result);
-                i--;
+        // Read the first numeric value, skip any further parsing if the string contains *only* one number.
+        ParsePosition pp = new ParsePosition(0);
```

Comment:
`parsePosition` or `parsePos`? pp is too abbreviated.

Reply:
Changed.